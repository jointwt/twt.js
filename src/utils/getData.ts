import fetch from 'cross-fetch';
import isResponseEmpty from './isResponseEmpty';

import {
  FetchOptions,
  UnknownResponse,
} from '../types.d';

export default async (options: FetchOptions): Promise<UnknownResponse> => {
  try {
    const {
      url,
      method,
      headers,
      body,
    } = options;

    const response = await fetch(url, {
      method,
      headers,
      ...body && { body: JSON.stringify(body) },
    });

    const emptyResponse = await isResponseEmpty(response);

    return {
      ok: response.ok,
      status: response.status,
      statusText: response.statusText,
      data: response.ok && !emptyResponse
        ? await response.json()
        : {},
    };
  } catch (error) {
    return {
      ok: false,
      status: -1,
      statusText: error.message,
      data: {},
    };
  }
};
