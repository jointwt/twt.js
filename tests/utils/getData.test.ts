import getData from '../../src/utils/getData';

it('fetches data from server when server returns a successful response', async () => {
  const data = await getData({
    url: 'https://successfulresponse.testurl',
    method: 'POST',
    body: {
      foo: 'bar',
    },
  });

  expect(data).toMatchSnapshot();
});

it('fetches data from server when server returns an empty response', async () => {
  const data = await getData({
    url: 'https://emptyresponse.testurl',
    method: 'POST',
    body: {
      foo: 'bar',
    },
  });

  expect(data).toMatchSnapshot();
});

it('handles errors gracefully', async () => {
  const data = await getData({
    url: 'https://nonexistingurl.testurl',
    method: 'POST',
    body: {
      foo: 'bar',
    },
  });

  expect(data).toMatchSnapshot();
});
