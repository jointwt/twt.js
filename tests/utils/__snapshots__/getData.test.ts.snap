// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches data from server when server returns a successful response 1`] = `
Object {
  "data": Object {
    "data": "mocked successful response",
  },
  "ok": true,
  "status": 200,
  "statusText": "ok",
}
`;

exports[`fetches data from server when server returns an empty response 1`] = `
Object {
  "data": Object {},
  "ok": true,
  "status": 200,
  "statusText": "ok but empty",
}
`;

exports[`handles errors gracefully 1`] = `
Object {
  "data": Object {},
  "ok": false,
  "status": -1,
  "statusText": "general mocked error",
}
`;
